setupTransitionPseudoElements doesn't update the view transition dynamic style sheet.

mattwoodrow opened the pull request:
#### 35cfdf3fdab6aa072eb6203618c0f7b6189aab66
<pre>
setupTransitionPseudoElements doesn&apos;t update the view transition dynamic style sheet.
<a href="https://bugs.webkit.org/show_bug.cgi?id=265225">https://bugs.webkit.org/show_bug.cgi?id=265225</a>
&lt;<a href="https://rdar.apple.com/118703381">rdar://118703381</a>&gt;

Reviewed by Tim Nguyen.

Refactors Resolver::setViewTransitionGroupStyle to take the pseduo element type,
and builds the appropriate selector, so that it can be used for all view-transition
pseudo elements.

Adds the generation of the view-transition dynamic style sheet styles to
ViewTransition::setupTransitionPseudoElements.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/mix-blend-mode-only-on-transition-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/paused-animation-at-end-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/style-inheritance-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/unset-and-initial-view-transition-name-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/view-transition-name-on-removed-element-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/window-resize-aborts-transition-expected.txt:
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::ViewTransition::setupDynamicStyleSheet):
(WebCore::ViewTransition::setupTransitionPseudoElements):
(WebCore::ViewTransition::copyElementBaseProperties):
(WebCore::ViewTransition::updatePseudoElementStyles):
* Source/WebCore/dom/ViewTransition.h:
* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::viewTransitionSelector):
(WebCore::Style::Resolver::setViewTransitionStyles):
(WebCore::Style::Resolver::setViewTransitionGroupStyles): Deleted.
* Source/WebCore/style/StyleResolver.h:

Canonical link: <a href="https://commits.webkit.org/275144@main">https://commits.webkit.org/275144@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6e66dd857f6c8a673f75f44f6b89547687a33aed

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40986 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19999 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43364 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43545 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37078 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/43293 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23009 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17330 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33947 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41560 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16926 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35314 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14574 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/14690 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/36307 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44867 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37181 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36619 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/40379 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/15801 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/12966 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/38749 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17420 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9197 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/17471 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/17064 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/style/StyleResolver.cpp:
```diff
@@ -726,24 +726,59 @@ std::optional<DynamicMediaQueryEvaluationChanges> Resolver::evaluateDynamicMedia
-    selectorList.append(MutableCSSSelector::parsePseudoClassSelector("root"_s, context));
-    auto groupSelector = MutableCSSSelector::parsePseudoElementSelector("view-transition-group"_s, context);
+    groupSelector->setValue(selectorName);
     groupSelector->setArgumentList({ { name } });
```
You don't want to set the argument list when the selector name is `view-transition`

nt1m commented on Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
```diff
@@ -153,6 +158,69 @@ void RenderTreeUpdater::ViewTransition::buildPseudoElementGroup(const AtomString
     if (viewTransitionGroup)
         m_updater.m_builder.attach(*documentElementRenderer.view().viewTransitionRoot(), WTFMove(viewTransitionGroup), beforeChild);
+
+    auto& resolver = document->styleScope().resolver();
```
Is this whole block directly linked to render tree construction? I would prefer putting this code after `setHasViewTransitionPseudoElementTree(true);` in `ViewTransition::setupTransitionPseudoElements()` if not. Otherwise this code will re-run when the group gets `display: none;` removed from its style.

nt1m commented on Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
```diff
@@ -153,6 +158,69 @@ void RenderTreeUpdater::ViewTransition::buildPseudoElementGroup(const AtomString
+        if (capturedElement && capturedElement->newElement)
+            list.append(CSSPrimitiveValue::create("-ua-mix-blend-mode-plus-lighter"_s));
+        Ref<CSSValueList> valueList = CSSValueList::createCommaSeparated(WTFMove(list));
+        Ref<MutableStyleProperties> props = MutableStyleProperties::create();
```
```suggestion
        Ref valueList = CSSValueList::createCommaSeparated(WTFMove(list));
        Ref props = MutableStyleProperties::create();
```

nt1m commented on Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
```diff
@@ -153,6 +158,69 @@ void RenderTreeUpdater::ViewTransition::buildPseudoElementGroup(const AtomString
+        if (capturedElement && capturedElement->oldImage)
+            list.append(CSSPrimitiveValue::create("-ua-mix-blend-mode-plus-lighter"_s));
+        Ref<CSSValueList> valueList = CSSValueList::createCommaSeparated(WTFMove(list));
+        Ref<MutableStyleProperties> props = MutableStyleProperties::create();
```
```suggestion
        Ref valueList = CSSValueList::createCommaSeparated(WTFMove(list));
        Ref props = MutableStyleProperties::create();
```

nt1m commented on Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
```diff
@@ -153,6 +158,69 @@ void RenderTreeUpdater::ViewTransition::buildPseudoElementGroup(const AtomString
+        return;
+
+    {
+        Ref<MutableStyleProperties> props = MutableStyleProperties::create();
```
```suggestion
        Ref props = MutableStyleProperties::create();
```

nt1m commented on Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
```diff
@@ -153,6 +158,69 @@ void RenderTreeUpdater::ViewTransition::buildPseudoElementGroup(const AtomString
+    }
+
+    {
+        Ref<CSSValueList> list = CSSValueList::createCommaSeparated(CSSPrimitiveValue::create(makeString("-ua-view-transition-group-anim-", name)));
```
```suggestion
        Ref list = CSSValueList::createCommaSeparated(CSSPrimitiveValue::create(makeString("-ua-view-transition-group-anim-"_s, name)));
```

nt1m commented on Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
```diff
@@ -153,6 +158,69 @@ void RenderTreeUpdater::ViewTransition::buildPseudoElementGroup(const AtomString
+
+    {
+        Ref<CSSValueList> list = CSSValueList::createCommaSeparated(CSSPrimitiveValue::create(makeString("-ua-view-transition-group-anim-", name)));
+        Ref<MutableStyleProperties> props = MutableStyleProperties::create();
```
```suggestion
        Ref props = MutableStyleProperties::create();
```

nt1m commented on Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
```diff
@@ -153,6 +158,69 @@ void RenderTreeUpdater::ViewTransition::buildPseudoElementGroup(const AtomString
+    if (!capturedElement->oldProperties)
+        return;
+
+    Ref<MutableStyleProperties> props = MutableStyleProperties::createEmpty();
```
```suggestion
    Ref props = MutableStyleProperties::createEmpty();
```

nt1m commented on Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
```diff
@@ -153,6 +158,69 @@ void RenderTreeUpdater::ViewTransition::buildPseudoElementGroup(const AtomString
+    Ref<StyleRuleKeyframe> keyframe = StyleRuleKeyframe::create(WTFMove(props));
+    keyframe->setKeyText("from"_s);
+
+    Ref<StyleRuleKeyframes> keyframes = StyleRuleKeyframes::create(AtomString(makeString("-ua-view-transition-group-anim-", name)));
```
```suggestion
    Ref keyframe = StyleRuleKeyframe::create(WTFMove(props));
    keyframe->setKeyText("from"_s);

    Ref keyframes = StyleRuleKeyframes::create(AtomString(makeString("-ua-view-transition-group-anim-"_s, name)));
```

nt1m commented on Source/WebCore/style/StyleResolver.cpp:
```diff
@@ -726,24 +726,56 @@ std::optional<DynamicMediaQueryEvaluationChanges> Resolver::evaluateDynamicMedia
+        break;
+    case CSSSelector::PseudoElement::ViewTransitionImagePair:
+        selectorName = "view-transition-image-pair"_s;
+            break;
```
nit: weird indentation

nt1m commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -318,11 +320,78 @@ ExceptionOr<void> ViewTransition::captureNewState()
     protectedDocument()->setHasViewTransitionPseudoElementTree(true);
-    // FIXME: Implement step 9.
+
+    for (auto& iter : m_namedElements.map())
```
```suggestion
    for (auto& [key, value] : m_namedElements.map())
```

Can we use these kinds of iterators?

mattwoodrow committed:
setupTransitionPseudoElements doesn't update the view transition dynamic style sheet.
https://bugs.webkit.org/show_bug.cgi?id=265225
<rdar://118703381>

Reviewed by Tim Nguyen.

Refactors Resolver::setViewTransitionGroupStyle to take the pseduo element type,
and builds the appropriate selector, so that it can be used for all view-transition
pseudo elements.

Adds the generation of the view-transition dynamic style sheet styles to
ViewTransition::setupTransitionPseudoElements.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/mix-blend-mode-only-on-transition-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/paused-animation-at-end-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/style-inheritance-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/unset-and-initial-view-transition-name-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/view-transition-name-on-removed-element-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/window-resize-aborts-transition-expected.txt:
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::ViewTransition::setupDynamicStyleSheet):
(WebCore::ViewTransition::setupTransitionPseudoElements):
(WebCore::ViewTransition::copyElementBaseProperties):
(WebCore::ViewTransition::updatePseudoElementStyles):
* Source/WebCore/dom/ViewTransition.h:
* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::viewTransitionSelector):
(WebCore::Style::Resolver::setViewTransitionStyles):
(WebCore::Style::Resolver::setViewTransitionGroupStyles): Deleted.
* Source/WebCore/style/StyleResolver.h:

Canonical link: https://commits.webkit.org/275144@main
